Migrates Orders list - part 1

sarjon opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migrates orders list.
| Type?         | refacto 
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Migrate "Orders > Orders" page #10582
| How to test?  | Should work the same as legacy list. :warning: Rebuild assets before testing.
<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/13988)
<!-- Reviewable:end -->

sarjon committed:
Adds controller, template and routing

sarjon committed:
Configures query builder, definition factory and grid factory for Orders

sarjon committed:
Adds new client, customer and total fields to order

sarjon committed:
Adds order status column to Order grid

sarjon committed:
Adds PDF & Date columns

sarjon committed:
Adds actions column

sarjon committed:
Adds company column

sarjon committed:
Adds delivery country column to grid

sarjon committed:
Adds LinkGroup column to orders list

sarjon committed:
Implements generating order invoice

sarjon committed:
Implements generating delivery slip pdf

sarjon committed:
Configures delivery slip generating ulr in orders grid

sarjon committed:
Adds filter inputs in grid

sarjon committed:
Adds countries column only when countries are used

sarjon committed:
Adds order status selection in filters form

sarjon committed:
Adds orders grid filtering

sarjon committed:
Fixes filtering

sarjon committed:
Adds JS extensions to Orders grid

sarjon committed:
Implements orders sorting

sarjon committed:
Adds reusable form submitting for bulk action

sarjon committed:
Adds reusable JS component to submit forms in modal

sarjon committed:
Implements changing orders state via bulk action

sarjon committed:
Fixes template of link group column

sarjon committed:
Adds filters reseting on grid

sarjon committed:
Adds exporting orders

sarjon committed:
Exclude company from export when B2B is disabled

sarjon committed:
Skips translating choices for order statuses

sarjon committed:
Aligns order price to the right

sarjon committed:
Adds KPIs for orders

sarjon committed:
Uses context date format to display order date

sarjon committed:
Adds security annotations

sarjon committed:
Applies CS fixer

sarjon committed:
Removes empty line

sarjon commented on src/Adapter/Order/CommandHandler/ChangeOrdersStatusHandler.php:
```diff
@@ -0,0 +1,97 @@
+                $templateVars['{followup}'] = str_replace('@', $order->shipping_number, $carrier->url);
+            }
+
+            if ($history->addWithemail(true, $templateVars)) {
```
@matks this one is very tricky, `$history->addWithemail()`can return `false` in two cases:

1. When `OrderHistory` was not created (which is OK)
2. When `OrderHistory` is created, but email was not sent to customer

I could workaround this by splitting this method into:

```php
if (!$history->add()) {
    // throw saving order exception
}

if (!$history->sendEmail()) {
    // throw email sending exception
}
```
Unfortunately, it does not solve issue completely. In legacy, if `$history->addWithemail()` fails, it continues to loop through other orders attempting to change their status and in the end if shows list of errors:

![Screenshot (1)](https://user-images.githubusercontent.com/15143151/58637225-ba448b80-82fa-11e9-9d56-1a84756ddeab.png)

I'm wondering how should we keep the same behavior when using command to change order status? Should we keep track of errors and throw exception with multiple errors? Wdyt?

matks replied:
The question can be converted in a more global matter: "When performing one bulk action, when one of the actions fail, should we 1) stop there and display the error message to user or should we 2) skip it, attempt to perform the action on the remaining items, and at the end display a list of the failed usecases ?"

So far we have followed the strategy 1) and I like this because one error might be a hint of something wrong happening, so it's good to wait for user to re-try to make sure the bulk action was not triggered by mistake.

However the decision is a Product Choice so ping @colinegin @marionf

sarjon replied:
Should we consider email sending failure an error? This happens pretty often, how do we inform user about it?

colinegin replied:
@MatShir this is related to orders listing, so I let you decide with @marionf ! ;)

marionf replied:
@matks I think it's better to skip it and perform action in the remaining items, and display the list of the failed usecases at the end
This way if some of remaining items are not failing, the action is done for these items

matks replied:
So it's decided 😉ping @sarjon 
And I'll create an issue to apply this strategy to all pages

matks replied:
https://github.com/PrestaShop/PrestaShop/issues/14179

sarjon commented on src/Adapter/PDF/DeliverySlipPdfGenerator.php:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function generatePDF(array $orderId)
```
I didn't want to introduce new interface for generating PDFs, so I've used existing one. However, I've used it in a way it wasn't meant to be implemented. :confused: You can take a look at `OrderController` to see how it's called.

Wdyt about it?

matks replied:
You couldn't use it because this Interface is not generic enough.
- the name is not right, even the phpDoc says it: "Generates PDF out of given object and template using legacy generator." => should be something like LegacyHtmlPdfGenerator
- the array `$objectCollection` should be type-hinted with something like HTMLTemplateInterface because when you look at the Pdf implementation you can see it only accepts HTMLTemplate objects

So I'm wondering whether we should deprecate it.
I think we can go for a new specific Interface for us. IMO it's like dealing with mail sending: there must be 2 interfaces:
- the MailSender itself, handling SMTP, IMAP, whatever technical matters required to send an email
- the Domain MailSender which has functions such as `sendRegistrationMail()`, `sendOrderApprovalMail()` and does the boilerplate work of getting the right template, putting the right recipient, the right title ...

So I think we can go for one DeliverySlipPdfWriter that relies on a PdfGenerator.

sarjon replied:
> So I'm wondering whether we should deprecate it.

This is a big problem imo. We cannot introduce new interface in 1.7.5 and then deprecate in 1.7.7, it's just wrong and I think you understand it. :/

How could we solve it? It's a recurrent issue we face.

Maybe each time we introduce new component, it should be done in separate PR, so everyone can discuss it and provide feedback, otherwise this kind of interfaces blurs in context of migration PRs and gets merged without being well designed. :/

matks replied:
> > So I'm wondering whether we should deprecate it.
> 
> This is a big problem imo. We cannot introduce new interface in 1.7.5 and then deprecate in 1.7.7, it's just wrong and I think you understand it. :/

I dont think @mickaelandrieu would be against it 😛

> Maybe each time we introduce new component, it should be done in separate PR, so everyone can discuss it and provide feedback, otherwise this kind of interfaces blurs in context of migration PRs and gets merged without being well designed. :/

Yes indeed

sarjon replied:
Since this could be considered as a separate feature, I suggest improving it with new implementation in separate PR, wdyt?

matks replied:
Agreed 👍

matks replied:
Added to issue todo-list ✅

sarjon committed:
Adds missing docblock

sarjon committed:
Removes duplicate filter config

sarjon commented on src/Core/Grid/Query/OrderQueryBuilder.php:
```diff
@@ -0,0 +1,275 @@
+
+        $dateComparisonFilters = [
+            'date_add' => 'o.`date_add`',
+        ];
```
@matks writing queries manually is repetitive along all grids. In addition, configuring filters for each grid is very painful and debugging to make sure they work well is time consuming whenever we do it.

It's also a copy/paste approach, since most of the time we copy code from another query builder and apply some changes. 

I think this could be a first step towards easier filters configuration. Wdyt?

It could be improved by extracting below `foreach` into separate service like:

```php
$filtersApplicator = new FiltersApplicator();

$config = [
    'date_filters' => [ ... ],
    'strict_filters' => [ ... ],
    ...
];

$filtersApplicator->apply($queryBuilder, $config);
```

Of course it's not well thought at this point, but this is something we could improve thus easing grids migration / creation.

@mickaelandrieu I dare to ping you on this as well. :stuck_out_tongue:

matks replied:
I think it's a good idea 👍. To make it easy and efficient to use I suggest we do the following:
- `FiltersApplicator` must provide an easy and simple API such as `addFilter()`, `addDateFilter()` ... it must aim for a Developer Experience similar to Symfony FormTypes
- we'll make a dedicated integration testsuite for it (probably with phpunit, behat will bring no value here) to make sure it is battle-tested because if we have a bug in it, it'll make QueryBuilders very hard to debug

sarjon replied:
Ok, but this should be done in separate PR I think. :)

matks replied:
Agreed, and I see you started it 😄 
We can merge this PR then refactor it 😉

sarjon commented on src/Core/Grid/Query/OrderQueryBuilder.php:
```diff
@@ -0,0 +1,275 @@
+            'customer' => 'customer',
+            'osname' => 'osl.name',
+            'date_add' => 'o.`date_add`',
+        ];
```
Same issue with sorting.

sarjon commented on src/Core/Grid/Definition/Factory/OrderGridDefinitionFactory.php:
```diff
@@ -0,0 +1,413 @@
+                                'route_param_name' => 'orderId',
+                                'route_param_field' => 'id_order',
+                            ],
+                        ],
```
@matks for this config, do you think it's worth introducing objects here?

matks replied:
For LinkGroupColumn options only ?

sarjon replied:
Yeah, is it worth?

matks replied:
No 😺

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Order/Blocks/change_orders_status_modal.html.twig:
```diff
@@ -0,0 +1,56 @@
+  'closable': true,
+  'actions': [{
+    'type': 'button',
+    'label': 'Update Order Status'|trans({}, 'Admin.Orderscustomers.Feature'),
```
What about _Update status_ instead? Do you agree that this will introduce a new behavior since it won't be necessary to go inside an order page to update its status?

sarjon replied:
I've just used the same wording as in legacy, if you think it should be changed, it's :+1: for me.

LouiseBonnard replied:
@sarjon, thanks for your quick reply! Am I right about the behavior or is it displayed in another context?

sarjon replied:
It's a legacy page, but the context is the same (change order status in bulk action) in new page as well, except it's displayed in modal. https://prnt.sc/nxujkw

LouiseBonnard replied:
Alright. Let's write _Update status_ in this case, localized in _Admin.Orderscustomers.Feature_ as well, please. :-)

sarjon replied:
Done. :+1:

sarjon replied:
Done. :+1:

sarjon committed:
Process all orders when updating status in bulk action

sarjon committed:
Rename ChangeOrdersStatus to BulkChangeOrderStatus

sarjon committed:
Update wording

matks commented on src/Adapter/PDF/DeliverySlipPdfGenerator.php:
```diff
@@ -0,0 +1,84 @@
+        $order = new Order((int) $orderId);
+
+        if (!Validate::isLoadedObject($order)) {
+            die($this->translator->trans(
```
We can at least replace this `die()` with an Exception 😉

matks commented on src/Adapter/PDF/OrderInvoicePdfGenerator.php:
```diff
@@ -0,0 +1,84 @@
+        $orderId = reset($orderId);
+        $order = new Order((int) $orderId);
+        if (!Validate::isLoadedObject($order)) {
+            die($this->translator->trans(
```
No `die()` on my watch 😄

matks commented on src/Core/Grid/Query/OrderQueryBuilder.php:
```diff
@@ -0,0 +1,274 @@
+            ->addSelect('CONCAT(LEFT(cu.`firstname`, 1), \'. \', cu.`lastname`) AS `customer`')
+            ->addSelect('os.color, o.payment')
+            ->addSelect('o.date_add, cu.company, cl.name AS country_name, o.invoice_number, o.delivery_number')
+            ->addSelect('IF ((' . $newCustomerSubSelect->getSQL() . ') > 0, 0, 1) AS new')
```
We'll have to check that this statement does not introduce performance issues. Subqueries can be terrible memory and time consumers

matks commented on src/Core/Util/ColorBrightnessCalculator.php:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @return float|int
+     */
+    private function calculate($hexColor)
```
I love this feature ^^

matks replied:
We can unit test that

sarjon replied:
Done.

matks commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -0,0 +1,267 @@
+     * Shows list of orders
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
+     *
```
AdminRestricted, no ?

sarjon replied:
Umm, what do you mean?

matks commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -0,0 +1,267 @@
+
+        // When using legacy generator,
+        // we want to be sure that displaying PDF is the last thing this controller will do
+        die;
```
```suggestion
        die();
```

sarjon replied:
Sure. :sweat_smile:

matks commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -0,0 +1,267 @@
+
+        // When using legacy generator,
+        // we want to be sure that displaying PDF is the last thing this controller will do
+        die;
```
```suggestion
        die();
```

sarjon replied:
Sure. :sweat_smile:

matks commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/OrderController.php:
```diff
@@ -0,0 +1,267 @@
+            );
+
+            $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+        } catch (ChangeOrderStatusException $e) {
```
Please now catch global Exception

sarjon replied:
Done, one `catch` below.

matks commented on src/PrestaShopBundle/Form/Admin/Sell/Order/ChangeOrdersStatusType.php:
```diff
@@ -0,0 +1,82 @@
+        ;
+
+        $builder->get('order_ids')
+            ->addModelTransformer(new CallbackTransformer(
```
Was that needed 😅 ?

sarjon replied:
Well, I don't need to do conversion in controller as form takes care of that. :stuck_out_tongue: 

It comes as `string[]` from UI, but I need `int[]`. :)

matks commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/orders/orders.yml:
```diff
@@ -0,0 +1,45 @@
+  methods: [POST]
+  defaults:
+    _controller: PrestaShopBundle:Admin/Sell/Order/Order:search
+    _legacy_controller: AdminOrders
```
Let's not forget _legacy_links

sarjon replied:
Yeah, but after full page migration only. :wink:

matks commented on src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml:
```diff
@@ -180,3 +180,20 @@ services:
+        - '@prestashop.core.grid.query_builder.order'
+        - '@prestashop.core.hook.dispatcher'
+        - '@prestashop.core.grid.query.doctrine_query_parser'
+        - '!php/const PrestaShop\PrestaShop\Core\Grid\Definition\Factory\OrderGridDefinitionFactory::GRID_ID'
```
We can stick to `order` for now but use constants all at once in a global PR

matks replied:
The PR could also provide constants for https://github.com/PrestaShop/PrestaShop/pull/13988/files#diff-5c353273f197707b8ac513ef4dbb4b2cR51

sarjon replied:
You mean have another PR that replaces hardcoded strings with constants everywhere?

matks commented on src/PrestaShopBundle/Resources/views/Admin/Common/Grid/Columns/Content/color.html.twig:
```diff
@@ -0,0 +1,30 @@
+ * International Registered Trademark & Property of PrestaShop SA
+ *#}
+
+{% set textColor = is_color_bright(record[column.options.color_field]) ? '#383838' : 'white' %}
```
👍

matks reviewed (changes requested):
Review done 😄 almost nothing to change, close to perfect 👏

sarjon committed:
Address review comments

sarjon committed:
Add unit test for ColorBrightnessCalculator

sarjon commented:
Review comments addressed.

matks commented:
@sarjon Approved ✅ can you rebase, then we got for QA ?

sarjon commented:
>  can you rebase, then we got for QA ?

done.

matks commented:
Rerun travis

sarahdib commented:
@sarjon 

there is some conflicting file can you resolve it before ?

sarjon commented:
@sarahdib rebased

sarahdib commented:
@sarjon waiting for @matks review :)

sarjon committed:
Fix pagination

sarjon committed:
Fix filtering orders by new client

sarjon commented:
@matks rebased PR, can you reapprove so we are ready for QA?

matks commented:
Here we go !

sarjon commented:
Travis needs to be restarted.

matks commented:
Let's do this, Travis !

sarjon commented:
Still failing. :sob:

matks commented:
Again ...

sarahdib commented:
Hello @sarjon 

- Missing column using multistore
When I'm using multistore the column store is display in legacy but not in migrated page.
![image](https://user-images.githubusercontent.com/3034677/63335038-ee997800-c33c-11e9-8b7b-6d529ecf1292.png)
![image](https://user-images.githubusercontent.com/3034677/63335043-f2c59580-c33c-11e9-84d0-24b1a3ba1768.png)

- The view action is not working. The action display orders list instead of the order (maybe it's not finish yet)

- The create order button is missing 

![image](https://user-images.githubusercontent.com/3034677/63335487-bc3c4a80-c33d-11e9-94a7-d870cacce776.png)
![image](https://user-images.githubusercontent.com/3034677/63335504-c3635880-c33d-11e9-8f72-862836f728fe.png)

- The listing is not correct in multishop. If I'm in the BO of shop 2 I do not have to see the order of shop 1
![image](https://user-images.githubusercontent.com/3034677/63336402-649ede80-c33f-11e9-9a40-8d7071960100.png)
![image](https://user-images.githubusercontent.com/3034677/63336419-6a94bf80-c33f-11e9-9a2b-c12b6cd81965.png)


- If I search all ID who contain "1" I found only the order 1 instead of 1, 10, 11 
![image](https://user-images.githubusercontent.com/3034677/63337737-cbbd9280-c341-11e9-9228-a60ab004a9d2.png)

![image](https://user-images.githubusercontent.com/3034677/63337761-d415cd80-c341-11e9-9d56-29dc7705cbd0.png)


- The search by date is not working
https://drive.google.com/file/d/1ZrWdo6nl5i4ki_aB4ib98GrYxlgqkR4L/view

sarjon committed:
Fix QA comments

sarjon committed:
Apply CS fixer

sarjon committed:
Builds assets

sarjon commented:
> Missing column using multistore

Fixed.

> The view action is not working. The action display orders list instead of the order (maybe it's not finish yet)

It's part of another PR.

> The create order button is missing

It's part of another PR.

> The listing is not correct in multishop. If I'm in the BO of shop 2 I do not have to see the order of shop 1

Fixed.

> If I search all ID who contain "1" I found only the order 1 instead of 1, 10, 11

Fixed

> The search by date is not working

Not related to this PR, it's a global issue being fixed another PR.

@sarahdib let me know if you have any other issues. :)

sarahdib commented:
Hello @sarjon 

Thanks for the fixes and precisions :)

I found a new bug 

- When I select all or some order and try to use bulk action to change the order status there is a 500 error (see the video)
If only one order is selected, there is no problem.
https://drive.google.com/file/d/17GB_iP-yfFiAHFawsdYgW9TO_463pGr5/view

- When I try to download the invoice there is an error it's ok for the delivery slip 
https://drive.google.com/file/d/15c0JxR1Ankgav2v3zBL5OdmQoNBWW1k-/view

- Sorry for the filter by id I saw with @colinegin because in another PR Julius tell me that on migrated pages if we search 3 we find 3 and not 13 or 23 anymore.  It was correct before. Sorry again.0

- Filter by new client return an error
https://drive.google.com/file/d/1HPz8qlqY_AnI86pt5uyv5Ejzelm__33k/view

sarjon commented:
Regarding 
> When I select all or some order and try to use bulk action to change the order status there is a 500 error (see the video)

and 

> When I try to download the invoice there is an error it's ok for the delivery slip

It works well for me. :confused: What's you PHP version?

sarjon committed:
Fix filtering by customer

sarjon commented:
ping @sarahdib

sarahdib commented:
@sarjon

Sorry I'm in the middle of the validation process for 1.7.6.1.

 I'm using php 7.1.19 google chrome browser.

sarahdib commented:
@sarjon  I don't know why the problem happen last time.
I re-install prestashop on my local and hosting test and it's ok for the invoice and the bulk action. Weird. I try several combinaison of multishop and it's ok now :) 

filter by new client -> OK
filter by id -> OK

Thank you @sarjon

sarjon commented:
> I re-install prestashop

In some cases that's all you need. :)

matks commented:
Thank you @sarjon and @sarahdib
